content/scripts/build.js

```diff
@@ -201,6 +232,7 @@ class Builder {
     this.logger = logger;
     this.selfHash = null;
     this.allTitles = null;
+    this.allRedirects = new Map();
```

Comment:
```suggestion
    this.allRedirects = null;
```
`Map`s are cool in that they contain a counter in their data structure. That's why you can do the idempotency check on the first couple of lines of `ensureAllRedirects` but a much simpler solution is to stick to objects. 
`null` means it hasn't fully run yet and `ensureAllRedirects()` needs to proceed. 
It's the trick being used in `ensureAllTitles()`. Doing a `Boolean(someObject)` is `O(1)` when all JS has to do is to see if `someObject` is a `null` or `{}`. 
Also, I quite like the consistency of `this.allTitles` and `this.allRedirects` being the same type.

Reply:
I think there's a stronger argument for the opposite approach, that is, re-implementing `this.allTitles` as a `Map`. The `O(1)` applies the same to both `Boolean(someObject)` as well as `someMap.size`, so in my view there are only downsides to using JS objects for large dictionaries, and `allTitles` is going to be on the order of 100K entries (assuming all locales). For example, https://github.com/mdn/stumptown-renderer/blob/727751fd14acc78fc44d45bd5b61d8236b6d88c5/content/scripts/build.js#L484 and https://github.com/mdn/stumptown-renderer/blob/727751fd14acc78fc44d45bd5b61d8236b6d88c5/content/scripts/build.js#L718 are really inefficient compared with the iterations that a `Map` provides. In both cases we're building a large temporary array and then garbage-collecting it soon after its use. So I agree we should be consistent in terms of the type of `this.allTitles` and `this.allRedirects`, but the type should be a `Map`. I would propose a separate task where we convert `this.allTitles` to a `Map` and also perhaps lowercase its keys. I think in general we should work with lowercase URI's and only the `mdn_url` in the metadata contains the "proper" casing.

Reply:
Yeah, those are great examples of where plain objects are bad. 
There's actually been [this](https://github.com/mdn/stumptown-renderer/blob/727751fd14acc78fc44d45bd5b61d8236b6d88c5/content/scripts/build.js#L427) for a long time and I was waiting for a strong reason to make the switch. 
The examples you pointed to are things that only really happen once per build.
And the `this.allTitles`, when it needs to be serialized into a JSON file, it needs to be object anyway. 

But I think we're in full agreement that:

1. `Map` is better than object in a whole host of ways
2. The `this.allTitles` and `this.allRedirects` should probably be consistent. 

So, how about this for a plan:

1. Merge this PR and accept that `this.allTitles` and `this.allRedirects` are of different types. 
2. A fresh mini-PR just to refactor `this.allTitles` to be a `Map` (I can own this if you own merging in from `master` into the `ks-renderer` branch)
3. Punt on the topic of lowercasing all the keys. (Not for long but just significantly distinct from this discussion about types)